File: solutions/968. Binary Tree Cameras.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int camera = 0;
public:
    
    int f(TreeNode* root) {
        if (!root) {
            return 1;
        }
        int lc = f(root->left) ;
        int rc = f(root->right) ;
        if (lc == -1 || rc == -1) {
            camera++;
            return 0;
        }
        if (lc == 0 || rc == 0) {
            return 1;
        }
        return -1;
    }
    
    int minCameraCover(TreeNode* root) {
        if (f(root) == -1) {
            camera++;
        }    
        // f(root);
        return camera;
    }
};

File: solutions/968. Binary Tree Cameras.java
​
class Solution {
int minimumCameras = 0;
    Set<TreeNode> covered = new HashSet<>(); // store the temp visited
    // Binary Tree Cameras | Live Coding with Explanation | Leetcode - 968
    // https://www.youtube.com/watch?v=2Gh5WPjAgJk
​
    public int minCameraCover(TreeNode root) {
        minimumCameras = 0;
        covered.add(null);
        minCameraCoverHelper(root, null);
        return minimumCameras;
    }
​
    public void minCameraCoverHelper(TreeNode root, TreeNode parent) {
        
        if (root == null) {
            return;
        }
​
        // faith
        minCameraCoverHelper(root.left, root);
        minCameraCoverHelper(root.right, root);
​
        // my work
        if (parent == null && !covered.contains(root)
                || !covered.contains(root.left)
                || !covered.contains(root.right)) {
            minimumCameras++;
            covered.add(parent);
            covered.add(root);
            covered.add(root.left);
            covered.add(root.right);
        }
    }
​
}
